gpbft/gpbft.go

```diff
@@ -1336,6 +1433,49 @@ func (c *convergeState) FindProposalFor(chain ECChain) (ConvergeValue, bool) {
 	return ConvergeValue{}, false
 }
 
+type broadcastState struct {
+	phaseAtLatestBroadcast Phase
+	roundAtLatestBroadcast uint64
+	messagesByRound        map[uint64][]*MessageBuilder
+}
+
+func newBroadcastState() *broadcastState {
+	return &broadcastState{
+		messagesByRound: make(map[uint64][]*MessageBuilder),
+	}
+}
+
+// record stores messages that are required should rebroadcast becomes necessary.
+// The messages stored depend on the current progress of instance. If the
+// instance progresses to DECIDE then only the decide message will be recorded.
+// Otherwise, all messages except QUALITY from the current and previous rounds
+// (i.e. the latest two rounds) are recorded.
+//
+// Note, the messages recorded are more than what FIL-00896 strictly requires at
+// the benefit of less reliance on rebroadcast and reduction in the possibility
+// of participants getting stuck.
+func (bs *broadcastState) record(round uint64, mb *MessageBuilder) {
+	bs.phaseAtLatestBroadcast = mb.payload.Step
+	bs.roundAtLatestBroadcast = round
```

Comment:
This `round` parameter raises questions since there is also `mb.payload.Round`, used below. 

In all the call paths into `broadcast` except the DECIDE phase, the `round` parameter comes from `i.round` and is written into `mb.payload.Round`, so the payload value could be used. For DECIDE, the parameter and payload is 0, but `i.round` is some larger value. The round will be ignored below, but I guess the reason for this is to interact with `progressedSinceLatestBroadcast`.

I think this is correct, but could be simpler. I consider `i.round` is rather meaningless once we reach the DECIDE phase. I'm hesitant to zero it since it's useful for introspection, but I think we should ignore it. How about encapsulating the complexity in `progressedSinceLatestBroadcast`? If the phase is DECIDE, return false without checking the round. Then here we can just record the payload values with no additional parameter. You might then rename the state members to be `lastBroadcastPhase|Round` to avoid confusion there too.

Reply:
That makes sense thanks Alex